File: core/flowfield/src/main/java/com/w2sv/flowfield/FlowField.java
package com.w2sv.flowfield;

import static processing.core.PConstants.TWO_PI;

import java.util.HashMap;
import java.util.Map;

import processing.core.PApplet;
import processing.core.PVector;

class FlowField {
    private static final float NOISE_ANGLE_SCALE = TWO_PI * 4; // 4 full rotations
    private final int granularity;
    private final float zOffIncrement;
    private float zOff = 0;

    private final Map<Long, Float> xNoiseCache = new HashMap<>();

    FlowField(int granularity, float zOffIncrement) {
        this.granularity = granularity;
        this.zOffIncrement = zOffIncrement;
    }

    void prepareFrame() {
        // Increment z-offset for temporal evolution of the field
        zOff += zOffIncrement;
    }

    float forceAngle(PVector pos, PApplet parent) {
        // Map particle position to grid cell coordinates
        int xCell = PApplet.floor(pos.x / granularity) + 1;
        int yCell = PApplet.floor(pos.y / granularity) + 1;
        // Pack cell coordinates into a long key for caching
        long key = packedLong(xCell, yCell);
        // Compute a unique x-offset for Perlin noise (cached per cell)
        float xNoise = xNoiseCache.computeIfAbsent(key, kk -> noiseXOffset(xCell, yCell, 0.1f));
        // Compute angle from noise
        return parent.noise(xNoise, 0, zOff) * NOISE_ANGLE_SCALE;
    }

    /**
     * Packs two ints into a single long (high 32 bits = a, low 32 bits = b).
     */
    private static long packedLong(int a, int b) {
        return (((long) a) << 32) | (b & 0xFFFFFFFFL);
    }

    /**
     * Maps a 2D grid cell (gx, gy) to a unique float suitable for Perlin noise input.
     *
     * <p>
     * Uses a pairing function (inspired by the Cantor pairing function) to ensure
     * each integer grid cell maps to a distinct value, while scaling it to keep
     * noise smooth.
     * </p>
     *
     * @param x     the x-index of the grid cell
     * @param y     the y-index of the grid cell
     * @param scale a scaling factor to control noise frequency (e.g., 0.1f)
     * @return a unique float representing the cell for use in noise()
     */
    private static float noiseXOffset(int x, int y, float scale) {
        return (((x + y) * (x + y + 1) / 2.0f + y) * scale);
    }
}

File: core/flowfield/src/main/kotlin/com/w2sv/flowfield/simulation/FlowField.kt
package com.w2sv.flowfield.simulation

import kotlin.math.PI
import kotlin.math.floor

/**
 * Generates a direction (angle in radians) for a given particle position.
 * Caches a precomputed noise-x offset per grid cell (sparse cache).
 */
internal class FlowField(
    private val cellsPerPixel: Float,
    private val timeIncrement: Float,
    private val noiseScale: Float = 0.1f,
    radianSamplingCoeff: Float
) {
    private val noiseToRadiansScale = 2f * PI.toFloat() * radianSamplingCoeff
    private var time = 0f

    private val xNoiseCache = HashMap<Long, Float>()

    /** advance temporal z slice (call once per frame) */
    fun prepareFrame() {
        time += timeIncrement
    }

    /** Returns an angle in radians for the given world position. */
    fun forceAngle(pos: Vec2): Float {
        // compute grid cell indices corresponding to pos
        val xCell = floor(pos.x * cellsPerPixel).toInt() + 1
        val yCell = floor(pos.y * cellsPerPixel).toInt() + 1
        val key = packedLong(xCell, yCell)

        val xNoise = xNoiseCache.getOrPut(key) { noiseXOffset(xCell, yCell, noiseScale) }
        return ProcessingNoise.noise(xNoise, 0f, time) * noiseToRadiansScale
    }

    private fun packedLong(a: Int, b: Int): Long =
        (a.toLong() shl 32) or (b.toLong() and 0xFFFFFFFFL)

    /**
     * Cantor-inspired pairing function scaled for noise input.
     * Ensures (x,y) map uniquely to a float and keeps values small for smooth noise.
     */
    private fun noiseXOffset(
        x: Int,
        y: Int,
        scale: Float
    ): Float {
        val s = (x + y).toFloat()
        val paired = (s * (s + 1f) * 0.5f) + y
        return paired * scale
    }
}
